README.md
# Simple and efficient serialize/deserialize library.

KTLV is acronym for Key-Type-Length-Value which mean
a method which used to encode and decode your data.

KTLV was inspired by [Apache Thrift](https://thrift.apache.org/).
Thrift is more mature and more wide used, but it's Erlang encoder
has a couple of disadvantages comparing with KTLV:

* it is 7 times slower in average;
* it produces a slightly bigger binary (near 8%);
* it's documentation is very, very poor and the source code is
 undocumented and contains no specifications.

On the other hand, KTLV provides only object encoding and
decoding features and it is not kinda protocol or transport
as Thrift is. Also there is no any
[Thrift IDL](https://thrift.apache.org/docs/idl) and IDL compiler
in KTLV. You should implement your own codec for the
programming language X by yourself instead.

The one of the most useful features KTLV taken from Thrift
is keyed elements, which can help to achieve backward and
forward compatibility between codecs.


debian/changelog
ktlv (2.3.1) unstable; urgency=low

  * Add missing HOME dir for go compiler
  * Make Golang package go mod compatible

 -- Aleksey Morarash <aleksey.morarash@gmail.com>  Thu, 26 Nov 2020 16:55:33 +0200

ktlv (2.2.2) unstable; urgency=low

  * Allow nil values for lists
  * Simplify Golang encoder

 -- Aleksey Morarash <aleksey.morarash@gmail.com>  Thu, 11 Apr 2019 20:47:42 +0300

ktlv (2.2.1) unstable; urgency=low

  * Adapt golang lib path to Debian Stretch rules
  * Fix copyright year
  * Specs: remove Vcs-Browser and Vcs-Git links
  * Upgrade Debian Policy version references

 -- Aleksey Morarash <aleksey.morarash@gmail.com>  Sat, 28 Apr 2018 11:36:43 +0300

ktlv (2.2.0) unstable; urgency=low

  * golang: extend API with Search*() functions

 -- Aleksey Morarash <aleksey.morarash@gmail.com>  Wed, 11 Apr 2018 14:04:43 +0300

ktlv (2.1.0) unstable; urgency=low

  * golang: add Dict.GetListOfUint64Def() method

 -- Aleksey Morarash <aleksey.morarash@gmail.com>  Thu, 16 Nov 2017 20:54:38 +0200

ktlv (2.0.0) unstable; urgency=low

  * golang: add DecodeElem() API function
  * golang: decoders return all successfully decoded elems on error
  * golang: refactor, stage 2
  * golang: remove redundant type Key
  * golang: remove redundant type FType
  * golang: refactor, remove deprecated API

 -- Aleksey Morarash <aleksey.morarash@gmail.com>  Fri, 03 Nov 2017 14:09:50 +0200

ktlv (1.2.0) unstable; urgency=low

  * golang: extend API

 -- Aleksey Morarash <aleksey.morarash@gmail.com>  Tue, 19 Sep 2017 17:33:38 +0300

ktlv (1.1.0) unstable; urgency=low

  * golang: add io.Writer-based encoding API

 -- Aleksey Morarash <aleksey.morarash@gmail.com>  Sun, 30 Jul 2017 22:54:26 +0300

ktlv (1.0.0) unstable; urgency=low

  * golang: add getters without default for some types
  * API change: rename getters with default values to *Def

 -- Aleksey Morarash <aleksey.morarash@gmail.com>  Wed, 21 Jun 2017 21:42:45 +0300

ktlv (0.9.0) unstable; urgency=low

  * golang: implement DataDict.GetUint16() method

 -- Aleksey Morarash <aleksey.morarash@gmail.com>  Mon, 29 May 2017 08:55:43 +0300

ktlv (0.8.1) unstable; urgency=low

  * adapt to Golang 1.7.4

 -- Aleksey Morarash <aleksey.morarash@gmail.com>  Tue, 21 Feb 2017 12:35:30 +0200

ktlv (0.8.0) unstable; urgency=low

  * golang: add some getters

 -- Aleksey Morarash <aleksey.morarash@gmail.com>  Thu, 12 Jan 2017 17:12:07 +0200

ktlv (0.7.0) unstable; urgency=low

  * golang: add new API without panics
  * gitignore *.swp files

 -- Aleksey Morarash <aleksey.morarash@gmail.com>  Tue, 20 Dec 2016 21:56:36 +0200

ktlv (0.6.0) unstable; urgency=low

  * golang: extend API with field getters

 -- Aleksey Morarash <aleksey.morarash@gmail.com>  Sat, 29 Oct 2016 14:23:24 +0300

ktlv (0.5.0) unstable; urgency=low

  * golang: extend API for with maps
  * golang makefile: add fmt target
  * golang makefile: remove unused code
  * extend gitignore file
  * golang: refactor type names
  * debianize Golang code
  * tests minor fix
  * implement codec for Go
  * license: year fix

 -- Aleksey Morarash <aleksey.morarash@gmail.com>  Mon, 12 Sep 2016 07:10:21 +0300

ktlv (0.4.0) unstable; urgency=low

  * erlang: add encd/1 API call

 -- Aleksey Morarash <aleksey.morarash@gmail.com>  Wed, 27 Apr 2016 14:51:35 +0300

ktlv (0.3.0) unstable; urgency=low

  * python: implement encoder

 -- Aleksey Morarash <aleksey.morarash@gmail.com>  Fri, 05 Feb 2016 17:14:12 +0200

ktlv (0.2.1) unstable; urgency=low

  * improve erlang-to-python test
  * python codec: fix bug in uint24 decoding
  * extend unit test for Erlang codec

 -- Aleksey Morarash <aleksey.morarash@gmail.com>  Tue, 17 Nov 2015 17:06:21 +0200

ktlv (0.2.0) unstable; urgency=low

  * add new types: int{8,16,24,32,64}, list_of_int{8,16,24,32,64}
  * dispose of magic numbers in unit tests

 -- Aleksey Morarash <aleksey.morarash@gmail.com>  Mon, 16 Nov 2015 17:06:08 +0200

ktlv (0.1.0) unstable; urgency=low

  * Initial release.

 -- Aleksey Morarash <aleksey.morarash@gmail.com>  Wed, 11 Nov 2015 14:34:00 +0200


golang/src/ktlv/ktlv.go
package ktlv

import "errors"

const (
	Bool   = 0
	Uint8  = 1
	Uint16 = 2
	Uint24 = 3
	Uint32 = 4
	Uint64 = 5
	Double = 6
	String = 7
	Bitmap = 8
	Int8   = 9
	Int16  = 10
	Int24  = 11
	Int32  = 12
	Int64  = 13

	List_of_String = 50
	List_of_Uint8  = 51
	List_of_Uint16 = 52
	List_of_Uint24 = 53
	List_of_Uint32 = 54
	List_of_Uint64 = 55
	List_of_Double = 56
	List_of_Int8   = 57
	List_of_Int16  = 58
	List_of_Int24  = 59
	List_of_Int32  = 60
	List_of_Int64  = 61

	Min_Int8   = int8(-0x80)
	Min_Int16  = int16(-0x8000)
	Min_Int24  = int32(-0x800000)
	Min_Int32  = int32(-0x80000000)
	Min_Int64  = int64(-0x8000000000000000)
	Max_Int8   = int8(0x7f)
	Max_Int16  = int16(0x7fff)
	Max_Int24  = int32(0x7fffff)
	Max_Int32  = int32(0x7fffffff)
	Max_Int64  = int64(0x7fffffffffffffff)
	Min_Uint8  = uint8(0)
	Min_Uint16 = uint16(0)
	Min_Uint24 = uint32(0)
	Min_Uint32 = uint32(0)
	Min_Uint64 = uint64(0)
	Max_Uint8  = uint8(0xff)
	Max_Uint16 = uint16(0xffff)
	Max_Uint24 = uint32(0xffffff)
	Max_Uint32 = uint32(0xffffffff)
	Max_Uint64 = uint64(0xffffffffffffffff)
)

var t2s = map[uint8]string{
	Bool:           "Bool",
	Uint8:          "Uint8",
	Uint16:         "Uint16",
	Uint24:         "Uint24",
	Uint32:         "Uint32",
	Uint64:         "Uint64",
	Double:         "Double",
	String:         "String",
	Bitmap:         "Bitmap",
	Int8:           "Int8",
	Int16:          "Int16",
	Int24:          "Int24",
	Int32:          "Int32",
	Int64:          "Int64",
	List_of_String: "List_of_String",
	List_of_Uint8:  "List_of_Uint8",
	List_of_Uint16: "List_of_Uint16",
	List_of_Uint24: "List_of_Uint24",
	List_of_Uint32: "List_of_Uint32",
	List_of_Uint64: "List_of_Uint64",
	List_of_Double: "List_of_Double",
	List_of_Int8:   "List_of_Int8",
	List_of_Int16:  "List_of_Int16",
	List_of_Int24:  "List_of_Int24",
	List_of_Int32:  "List_of_Int32",
	List_of_Int64:  "List_of_Int64",
}

func FTypeToString(t uint8) string {
	return t2s[t]
}

var (
	ElementNotFound     = errors.New("no such element")
	TypeAssertionFailed = errors.New("unexpected element type")
)
